### PR TITLE
fix(reactive-controllers): update IS_MOBILE condition for MatchMediaController

### PR DESCRIPTION
**Pull request by TarunAdobe**
## Description

MatchMediaController would now `window.screen.availHeight` and `window.screen.availWidth` in conjunction with`window.matchMedia` to detect a mobile device accurately. 

## Related issue(s)

#4647 

## Motivation and context

Inconsistency in Picker, Action-Menu opening as `popover` vs `tray` in iPad devices after changing orientation.

## How has this been tested?

-   [ ] _Test case 1_
    1. Go here
    2. Do this
-   [ ] _Test case 2_

    1. Go here
    2. Do this

-   [ ] Did it pass in Desktop?
-   [ ] Did it pass in Mobile?
-   [ ] Did it pass in iPad?

## Types of changes

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)
-   [ ] Chore (minor updates related to the tooling or maintenance of the repository, does not impact compiled assets)

## Checklist

-   [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
-   [x] My code follows the code style of this project.
-   [ ] If my change required a change to the documentation, I have updated the documentation in this pull request.
-   [x] I have read the **[CONTRIBUTING](<(https://github.com/adobe/spectrum-web-components/blob/main/CONTRIBUTING.md)>)** document.
-   [x] I have added tests to cover my changes.
-   [ ] All new and existing tests passed.
-   [x] I have reviewed at the Accessibility Practices for this feature, see: [Aria Practices](https://www.w3.org/TR/wai-aria-practices/)

## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.

### CODE REVIEW

**Review by rubencarvalho — CHANGES_REQUESTED** [review]
Let us dive deeper into avoiding attaching an event listener to each component, i.e., let's explore the singleton approach.

File: tools/reactive-controllers/src/MatchMedia.ts [hunk 1]
```diff
@@ -27,23 +27,45 @@ export class MatchMediaController implements ReactiveController {
...
+        if (this.mediaMatches === event.matches) return;
+        this.mediaMatches = event.matches;
+        this.host.requestUpdate(this.key, !this.mediaMatches);
+    }
+
+    protected updateMatches(): void {
+        // check for window.screen.availWidth and window.screen.availHeight to confirm that the device is a mobile device
+        const newMatches =
+            this.mediaMatches &&
+            (window.screen.availWidth < 768 || window.screen.availHeight < 768);
```
Comment (lazd) [inline tools/reactive-controllers/src/MatchMedia.ts:65]: Very interesting! Wondering how this works around issues with `mediaMatches`?

**Review by lazd — COMMENTED** [review]
Agreed that this needs to be done as a singleton, if this controller is on every element, we don't want thousands of listeners.

File: tools/reactive-controllers/src/MatchMedia.ts [hunk 2]
```diff
@@ -15,35 +15,153 @@ export const DARK_MODE = '(prefers-color-scheme: dark)';
...
-    protected media: MediaQueryList;
+    protected isMobileScreenQuery!: MediaQueryList;
+    protected smallScreenQuery!: MediaQueryList;
+    protected largeScreenQuery!: MediaQueryList;
+    protected orientationQuery!: MediaQueryList;
+
+    protected mediaQueryList!: MediaQueryList;
 
-    constructor(host: ReactiveElement, query: string) {
+    protected screenSize: 'small' | 'large' = 'small';
```
Comment (rubencarvalho) [inline tools/reactive-controllers/src/MatchMedia.ts:37]: Could we create some `enum`s for `screenSize` and `orientation`?

File: tools/reactive-controllers/src/MatchMedia.ts [hunk 3]
```diff
@@ -15,35 +15,153 @@ export const DARK_MODE = '(prefers-color-scheme: dark)';
...
-        this.matches = this.media.matches;
-        this.onChange = this.onChange.bind(this);
+
+        // if the query is provided then we use it to determine the matches
+        if (query) {
+            this.mediaQueryList = window.matchMedia(query);
+            this.matches = this.mediaQueryList.matches;
+            this.mediaQueryList.addEventListener(
+                'change',
+                this.onChange.bind(this)
```
Comment (rubencarvalho) [inline tools/reactive-controllers/src/MatchMedia.ts:50]: nit: I believe our code style typically uses arrow functions for binding `this`

File: tools/reactive-controllers/src/MatchMedia.ts [hunk 4]
```diff
@@ -15,35 +15,153 @@ export const DARK_MODE = '(prefers-color-scheme: dark)';
...
 
     protected host: ReactiveElement;
 
-    protected media: MediaQueryList;
+    protected isMobileScreenQuery!: MediaQueryList;
+    protected smallScreenQuery!: MediaQueryList;
+    protected largeScreenQuery!: MediaQueryList;
+    protected orientationQuery!: MediaQueryList;
+
+    protected mediaQueryList!: MediaQueryList;
```
Comment (rubencarvalho) [inline tools/reactive-controllers/src/MatchMedia.ts:35]: nit: could we have a more descriptive name for this one? The name `mediaQueryList` confuses me a bit when reading through the code. Maybe something like `customMediaQueryList` would be clearer that this is used when a custom media query is passed?

File: tools/reactive-controllers/src/MatchMedia.ts [hunk 5]
```diff
@@ -15,35 +15,153 @@ export const DARK_MODE = '(prefers-color-scheme: dark)';
 export const IS_MOBILE =
     '(max-width: 700px) and (hover: none) and (pointer: coarse), (max-height: 700px) and (hover: none) and (pointer: coarse)';
 
+export const MOBILE_SCREEN_QUERY = '(hover: none) and (pointer: coarse)';
+export const SMALL_SCREEN_QUERY = '(max-width: 743px)';
+export const LARGE_SCREEN_QUERY = '(min-width: 744px)'; // iPad Mini 6 max dimension, want it to be a 'tablet' using Desktop UI
+export const ORIENTATION_QUERY = '(orientation:portrait)';
+
 export class MatchMediaController implements ReactiveController {
     key = Symbol('match-media-key');
```
Comment (rubencarvalho) [inline tools/reactive-controllers/src/MatchMedia.ts:24]: What is this `key` used for?

File: tools/reactive-controllers/src/MatchMedia.ts [hunk 6]
```diff
@@ -15,35 +15,153 @@ export const DARK_MODE = '(prefers-color-scheme: dark)';
...
-        this.host.requestUpdate(this.key, !this.matches);
+    protected onChange(): void {
+        // if the query is provided
+        if (this.mediaQueryList) {
+            this.matches = this.mediaQueryList.matches;
+            this.host.requestUpdate();
+            return;
+        }
+
+        const screenSize = this._determineScreenSize();
```
Comment (rubencarvalho) [inline tools/reactive-controllers/src/MatchMedia.ts:151]: Could we encapsulate this "check if is mobile" logic into a dedicated method?

```ts
private _checkIsMobile(): boolean {
  const screenSize = this._determineScreenSize();
  const orientation = this._determineScreenOrientation();
  const isMobileScreen = this.isMobileScreenQuery.matches;

  // if the screen size is small or the orientation is landscape and the device is mobile
  return isMobileScreen && (screenSize === screenSize.small || orientation === orientation.landscape);
}
```

and then

```ts
this.matches = _checkIsMobile()
```

**Review by Rajdeepc — APPROVED** [review]
This is good to unblock for now! Let's think through a context level approach in achieving this in a follow up PR!

**Pull request merged**